src/main/java/net/minecraftforge/items/wrapper/ShulkerItemStackInvWrapper.java

```diff
@@ -0,0 +1,247 @@
+/*
+ * Copyright (c) Forge Development LLC and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.minecraftforge.items.wrapper;
+
+import net.minecraft.core.Direction;
+import net.minecraft.core.NonNullList;
+import net.minecraft.nbt.CompoundTag;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.world.ContainerHelper;
+import net.minecraft.world.item.BlockItem;
+import net.minecraft.world.item.Item;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.Items;
+import net.minecraft.world.level.block.entity.BlockEntityType;
+import net.minecraftforge.common.capabilities.Capability;
+import net.minecraftforge.common.capabilities.ICapabilityProvider;
+import net.minecraftforge.common.util.LazyOptional;
+import net.minecraftforge.event.AttachCapabilitiesEvent;
+import net.minecraftforge.eventbus.api.SubscribeEvent;
+import net.minecraftforge.fml.common.Mod;
+import net.minecraftforge.items.CapabilityItemHandler;
+import net.minecraftforge.items.IItemHandler;
+import net.minecraftforge.items.ItemHandlerHelper;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Set;
+
+@Mod.EventBusSubscriber(modid = "forge")
+public class ShulkerItemStackInvWrapper implements IItemHandler, ICapabilityProvider
+{
+    private final ItemStack stack;
+    private final LazyOptional<IItemHandler> holder = LazyOptional.of(() -> this);
+
+    public ShulkerItemStackInvWrapper(ItemStack stack)
+    {
+        this.stack = stack;
+    }
+
+    @Override
+    public int getSlots()
+    {
+        return 27;
+    }
+
+    @Override
+    @NotNull
+    public ItemStack getStackInSlot(int slot)
+    {
+        if (slot < getSlots())
+        {
+            return getItemList().get(slot);
+        }
+        return ItemStack.EMPTY;
+    }
+
+    @Override
+    @NotNull
+    public ItemStack insertItem(int slot, @NotNull ItemStack stack, boolean simulate)
```

Comment:
The logic here seems significantly more complex than the equivalent in `ItemStackHandler`. Why is that?

Reply:
This logic was copied from InvWrapper; it may be a sign that InvWrapper needs a refactor. I'll look at using the logic from ItemStackHandler instead.

Reply:
(Fixed by [4076409](https://github.com/MinecraftForge/MinecraftForge/pull/8827/commits/40764098c55531042d61406596c758b5a659664e))